### PR TITLE
feat: replay WAL entries respect index

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
#4385
## What's changed and what's your intention?

replay WAL entries respect index

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/store-api/src/logstore.rs [hunk 1]
```diff
@@ -30,6 +30,21 @@ pub use crate::logstore::entry::Id as EntryId;
 use crate::logstore::provider::Provider;
 use crate::storage::RegionId;
 
+#[derive(Debug, Clone, Copy)]
+pub struct WalIndex {
+    pub region_id: RegionId,
+    pub from_peer_id: u64,
```
Comment (fengjiachun) [inline src/store-api/src/logstore.rs:36]: Can we just use `path` of index?

File: src/log-store/src/kafka/index/collector.rs [hunk 2]
```diff
@@ -148,16 +158,69 @@ impl GlobalIndexCollector {
...
+        &self,
+        datanode_id: u64,
+        provider: &KafkaProvider,
+        region_id: RegionId,
+        entry_id: EntryId,
+    ) -> Result<Option<(BTreeSet<EntryId>, EntryId)>> {
+        let path = default_index_file(datanode_id);
+        let exists = self
+            .operator
+            .is_exist(&path)
```
Comment (fengjiachun) [inline src/log-store/src/kafka/index/collector.rs:196]: How about `read` the data directly? This way, there will be one less call.

File: src/store-api/src/logstore.rs (hunk 1 above)
Comment (WenyXu) [reply src/store-api/src/logstore.rs]: Rename it to `datanode_id`.

File: src/log-store/src/kafka/index/collector.rs (hunk 2 above)
Comment (WenyXu) [reply src/log-store/src/kafka/index/collector.rs]: Resolved in [6a65cbf](https://github.com/GreptimeTeam/greptimedb/pull/4565/commits/6a65cbf5fab7658196c4cdba4ebcbecbcff38f92)

File: src/store-api/src/logstore.rs (hunk 1 above)
Comment (fengjiachun) [reply src/store-api/src/logstore.rs]: location_id is better

File: src/log-store/src/kafka/index.rs [hunk 3]
```diff
@@ -20,10 +20,11 @@ pub use collector::GlobalIndexCollector;
...
-    RegionWalVecIndex,
+    build_region_wal_index_iterator, NextBatchHint, RegionWalIndexIterator, MIN_BATCH_WINDOW_SIZE,
 };
+#[cfg(test)]
+pub(crate) use iterator::{MultipleRegionWalIndexIterator, RegionWalRange, RegionWalVecIndex};
 
-pub fn default_index_file(datanode_id: u64) -> String {
-    format!("__datanode/{datanode_id}/index.json")
+pub fn default_index_file(location_id: u64) -> String {
+    format!("__datanode/{location_id}/index.json")
```
Comment (fengjiachun) [inline src/log-store/src/kafka/index.rs:29]: __wal/{location_id}/index.json ?

File: src/mito2/src/wal.rs [hunk 4]
```diff
@@ -102,15 +102,24 @@ impl<S: LogStore> Wal<S> {
...
         match provider {
             Provider::RaftEngine(_) => Box::new(LogStoreEntryReader::new(
                 LogStoreRawEntryReader::new(self.store.clone()),
             )),
-            Provider::Kafka(_) => Box::new(LogStoreEntryReader::new(RegionRawEntryReader::new(
-                LogStoreRawEntryReader::new(self.store.clone()),
-                region_id,
-            ))),
+            Provider::Kafka(_) => {
+                let reader = if let Some(from_peer_id) = location_id {
```
Comment (fengjiachun) [inline src/mito2/src/wal.rs:112]:
```suggestion
                let reader = if let Some(location_id) = location_id {
```

File: src/mito2/src/wal.rs [hunk 5]
```diff
@@ -102,15 +102,24 @@ impl<S: LogStore> Wal<S> {
...
+                    LogStoreRawEntryReader::new(self.store.clone())
+                        .with_wal_index(WalIndex::new(region_id, from_peer_id))
```
Comment (fengjiachun) [inline src/mito2/src/wal.rs:114]:
```suggestion
                        .with_wal_index(WalIndex::new(region_id, location_id))
```

**Review by fengjiachun — APPROVED** [review]

File: src/log-store/src/kafka/index/collector.rs [hunk 6]
```diff
@@ -148,16 +159,65 @@ impl GlobalIndexCollector {
...
+                    return Ok(None);
+                } else {
+                    return Err(err).context(error::ReadIndexSnafu { path });
+                }
+            }
+        };
+
+        match DatanodeWalIndexes::decode(&bytes)?.provider(provider) {
+            Some(indexes) => {
+                let last_index = indexes.last_index();
```
Comment (discord9) [inline src/log-store/src/kafka/index/collector.rs:209]: should there be a check for providing impossible entry id(like entry id > last_index) which is unexpected and should report a error(and a corresponding unit test?)

**Comment by discord9** [comment]
other LGTM

**Review by discord9 — APPROVED** [review]

**Pull request merged**